teps/0122-reproducibility-of-complete-build-instructions.md

```diff
@@ -0,0 +1,279 @@
+---
+status: propoased
+title: Reproducibility of Complete Build Instructions
+creation-date: '2022-09-14'
+last-updated: '2022-09-14'
+authors:
+- '@chitrangpatel'
+see-also:
+---
+
+# TEP-0122: Reproducibility of Complete Build Instructions
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+- [Out-of-scope](#out-of-scope)
+- [Future Work](#future-work)
+<!-- /toc -->
+
+## Summary
+This proposal outlines what information is required in the [provenance](#provenance-for-executed-taskrun)to reproduce complete build instructions for taskruns. It also suggests where in the `SLSA v0.2` [provenance](https://slsa.dev/provenance/v0.2) to store this information. 
+ 
+## Motivation
+
+[Tekton Chains will report provenance as reproducible if a specific annotation is included in the TaskRun](https://docs.google.com/document/d/12yYP0M_zz5Tc9ftGnLwhDZ_tdZBq4mEod70ngWtHmLs/edit#bookmark=id.1sm6isih1d0n) and it includes Task steps and parameter values in the provenance it generates, however there is a great deal of Task spec contents it ignores (e.g. sidecar definitions) and runtime information it does not include (e.g. workspaces provided at runtime). 
+
+For example the following provenance snippet was generated by Tekton Chains ([v0.8.0](https://github.com/tektoncd/chains/releases/tag/v0.8.0)) from [this chains example Task](https://github.com/tektoncd/chains/blob/main/examples/kaniko/kaniko.yaml) - it contains step information (including the shas of the images run) but is missing other information such as the workspaces and the volumes that backed them:
```

Comment:
I wonder how to capture enough information about workspaces and volumes that helps reproducibility. A volume name, for example, does not provide any useful information about its contents.

Reply:
One way might be to capture the complete spec (if possible) and not have to pick and choose. That would also help us transparently update in case the spec is updated in the future.

Reply:
Workspaces are a bit interesting. For example, if a workspace is a PVC, that contained some preinstalled binary/ other data that the task uses, then we currently cannot guarantee reproducibility of the build because that could be unavailable later on. 

Additionally, what do we do if a workspace is a secret? Do we put the secret data in the provenance? (I don;t think so).

Reply:
At this point, we will only store the workspace information in the spec and that provided at runtime in the task run.